### PR TITLE
Implements Go e2e tests

### PR DESCRIPTION
**Pull request by M4tteoP**
Implements e2e tests porting to Go and refactoring the ones used by `coraza-proxy-wasm` and  `coraza-caddy`. It uniforms the tests and provides a simplified way to run them against any Coraza deployment

Output example against coraza-proxy-wasm running with Envoy:
```
▶ go run . --nulled-body
[1/3] Running health check: Health check
[Wait] Waiting for http://localhost:8081/status/200. Timeout: 15s
[Ok] Check successful, got status code 200
[2/3] Running health check: Proxy check
[Wait] Waiting for http://localhost:8080. Timeout: 15s
[Ok] Check successful, got status code 200
[3/3] Running health check: Header check
[Wait] Waiting for http://localhost:8080. Timeout: 15s
[Ok] Check successful, got status code 424
[1/9] Running test: Legit request
[Ok] Got status code 200, expected 200
[2/9] Running test: Denied request by URL
[Ok] Got status code 403, expected 403
[3/9] Running test: Legit request with legit body
[Ok] Got status code 200, expected 200
[4/9] Running test: Denied request with a malicious request body
[Ok] Got status code 403, expected 403
[5/9] Running test: Denied request with a malicious response header
[Ok] Got status code 403, expected 403
[6/9] Running test: Denied request with a malicious response body
[Ok] Response body filled with null bytes
[Ok] Got status code 200, expected 403
[7/9] Running test: Denied request with XSS query parameters
[Ok] Got status code 403, expected 403
[8/9] Running test: Denied request with SQLi query parameters
[Ok] Got status code 403, expected 403
[9/9] Running test: CRS rule 913100 sending malicious UA
[Ok] Got status code 403, expected 403
```

### CODE REVIEW

File: http/e2e/e2e.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+// Custom rule for Coraza config check (ensuring that these configs are used)
+SecRule &REQUEST_HEADERS:coraza-e2e "@eq 0" "id:100,phase:1,deny,status:424,msg:'Coraza E2E - Missing header'"
+// Custom rules for e2e testing
+SecRule REQUEST_URI "@streq /admin" "id:101,phase:1,t:lowercase,deny"
+SecRule REQUEST_BODY "@rx maliciouspayload" "id:102,phase:2,t:lowercase,deny"
+SecRule RESPONSE_HEADERS::status "@rx 406" "id:103,phase:3,t:lowercase,deny"
+SecRule RESPONSE_BODY "@contains responsebodycode" "id:104,phase:4,t:lowercase,deny"
+*/
+
+const configCheckStatusCode = 424
```
Comment (jcchavezs) [inline http/e2e/e2e.go:36]: https://http.cat/424

File: http/e2e/e2e.go [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+const healthCheckTimeout = 15 // Seconds
+
+func main() {
+	// Initialize variables
+	corazaEnvoy := false
+	corazaEnvoyString := os.Getenv("CORAZA_ENVOY")
```
Comment (jcchavezs) [inline http/e2e/e2e.go:42]: Not sure we should use env vars or flags.

File: http/e2e/e2e.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+			expectedCode: 200,
+		},
+		{
+			name:         "Coraza config check",
+			url:          urlUnfiltered,
+			expectedCode: configCheckStatusCode,
+		},
+	}
+
+	tests := []struct {
```
Comment (jcchavezs) [inline http/e2e/e2e.go:83]: We could move the tests outside the function.

File: http/e2e/e2e.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+			httpMethod:   "GET",
+			expectedCode: 403,
+		},
+	}
+
+	// Check health endpoint
+	totalChecks := len(healthChecks)
+	for currentCheckIndex, healthCheck := range healthChecks {
+		fmt.Printf("[%d/%d]Running health check: %s\n", currentCheckIndex+1, totalChecks, healthCheck.name)
+		client := http.Client{}
```
Comment (jcchavezs) [inline http/e2e/e2e.go:175]: We could reuse the clients, no need to create a new one every time.

File: http/e2e/e2e.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+		timeout := healthCheckTimeout
+		tick := time.Tick(time.Second)
+
+		for range tick {
+			req, _ := http.NewRequest(http.MethodGet, healthCheck.url, nil)
```
Comment (jcchavezs) [inline http/e2e/e2e.go:180]: We can reuse the requests also.

File: http/e2e/e2e.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+			if healthCheck.expectedCode != configCheckStatusCode {
+				//  The default e2e header is not added if we are checking that the expected config is loaded
+				req.Header.Add("coraza-e2e", "ok")
+			}
+			resp, err := client.Do(req)
```
Comment (jcchavezs) [inline http/e2e/e2e.go:185]: Don't forget to close the body afterward.

File: http/e2e/e2e.go [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+)
+
+// External configurable variables:
+// - CORAZA_ENVOY: Interruptions at response body phase are allowed to return 200 (Instead of 403), but with a body full of null bytes. Defaults to "false".
```
Comment (jcchavezs) [inline http/e2e/e2e.go:16]: I would call it `NULLED_BODY`, I would avoid to make the tests specific to ENVOY.

File: http/e2e/e2e.go (hunk 3 above)
Comment (M4tteoP) [reply http/e2e/e2e.go:83]: They are relying on variables that are created at runtime, moving the tests outside the function would require then some kind of translation with the right values 🧐

**Comment by jcchavezs** [comment]
Before we merge this we need to use this e2e in the http connector.

File: http/middleware_test.go [hunk 8]
```diff
@@ -402,6 +402,53 @@ func TestHttpServerWithRuleEngineOff(t *testing.T) {
 	}
 }
 
+func TestHttpServerWithE2e(t *testing.T) {
```
Comment (jcchavezs) [inline http/middleware_test.go:405]: Is this needed I don't see a direct connection with it.

Comment (M4tteoP) [reply http/middleware_test.go]: Thanks, it was a leftover

File: http/e2e/pkg/runner.go [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+		if test.requestBody != "" {
+			requestBody = strings.NewReader(test.requestBody)
+		}
+
+		req, err := http.NewRequest(test.requestMethod, test.requestURL, requestBody)
+		if err != nil {
+			return fmt.Errorf("could not make http request: %v", err)
+		}
+
+		if test.requestMethod == "POST" {
```
[the remark this answers is not in the document]
Comment (M4tteoP) [reply http/e2e/pkg/runner.go]: this logic was kind of redundant and led to malformed requests caught by Envoy. I removed it and moved the explanation where we add that content-type inside each test case

**Comment by jcchavezs** [comment]
Merging as I need to test it with another repo.

**Pull request merged**